Sources/JWTKit/RSA/SubjectPublicKeyInfo.swift

```diff
@@ -0,0 +1,106 @@
+    }
+
+    var algorithmIdentifier: RFC5480AlgorithmIdentifier
+
+    var key: ASN1BitString
+
+    init(derEncoded rootNode: ASN1Node, withIdentifier identifier: ASN1Identifier) throws {
+        // The SPKI block looks like this:
+        //
+        // SubjectPublicKeyInfo  ::=  SEQUENCE  {
+        //   algorithm         AlgorithmIdentifier,
+        //   subjectPublicKey  BIT STRING
+        // }
+        self = try DER.sequence(rootNode, identifier: identifier) { nodes in
+            let algorithmIdentifier = try RFC5480AlgorithmIdentifier(derEncoded: &nodes)
+            let key = try ASN1BitString(derEncoded: &nodes)
+
+            return SubjectPublicKeyInfo(algorithmIdentifier: algorithmIdentifier, key: key)
+        }
+    }
+
+    private init(algorithmIdentifier: RFC5480AlgorithmIdentifier, key: ASN1BitString) {
+        self.algorithmIdentifier = algorithmIdentifier
+        self.key = key
+    }
+
+    init(algorithmIdentifier: RFC5480AlgorithmIdentifier, key: [UInt8]) {
+        self.algorithmIdentifier = algorithmIdentifier
+        self.key = ASN1BitString(bytes: key[...])
+    }
+
+    func serialize(into coder: inout DER.Serializer, withIdentifier identifier: ASN1Identifier) throws {
+        try coder.appendConstructedNode(identifier: identifier) { coder in
+            try coder.serialize(self.algorithmIdentifier)
+            try coder.serialize(self.key)
+        }
+    }
+}
+
+struct RFC5480AlgorithmIdentifier: DERImplicitlyTaggable, Hashable {
+    static var defaultIdentifier: ASN1Identifier {
+        .sequence
+    }
+
+    var algorithm: ASN1ObjectIdentifier
+
+    var parameters: ASN1Any?
+
+    init(algorithm: ASN1ObjectIdentifier, parameters: ASN1Any? = nil) {
+        self.algorithm = algorithm
+        self.parameters = parameters
+    }
+
+    init(derEncoded rootNode: ASN1Node, withIdentifier identifier: ASN1Identifier) throws {
+        self = try DER.sequence(rootNode, identifier: identifier) { nodes in
+            let algorithmOID = try ASN1ObjectIdentifier(derEncoded: &nodes)
+
+            let parameters = nodes.next().map { ASN1Any(derEncoded: $0) }
+
+            return .init(algorithm: algorithmOID, parameters: parameters)
+        }
+    }
+
+    func serialize(into coder: inout DER.Serializer, withIdentifier identifier: ASN1Identifier) throws {
+        try coder.appendConstructedNode(identifier: identifier) { coder in
+            try coder.serialize(self.algorithm)
+            if let parameters = self.parameters {
+                try coder.serialize(parameters)
+            }
+        }
+    }
+}
+
+// MARK: Algorithm Identifier Statics
+
+extension RFC5480AlgorithmIdentifier {
+    static let rsa256 = RFC5480AlgorithmIdentifier(
+        algorithm: .AlgorithmIdentifier.sha256WithRSAEncryption
+    )
+
+    static let rsa384 = RFC5480AlgorithmIdentifier(
+        algorithm: .AlgorithmIdentifier.sha384WithRSAEncryption
+    )
+
+    static let rsa521 = RFC5480AlgorithmIdentifier(
```

Comment:
521 😱

Reply:
Not a typo.

Reply:
What the heck 😶

Reply:
A somewhat bizarre mathematical quirk of the NIST curves that were chosen. See https://crypto.stackexchange.com/questions/6219/why-do-the-elliptic-curves-recommended-by-nist-use-521-bits-rather-than-512 for some gory details.

Reply:
Actually that _is_ a typo, RSA is not based on elliptic curves. It should be 512 as that's the hashing algorithm

Reply:
Damn, you're right. Nice catch! (And doubly so with me incorrectly arguing the opposite 😅)